PIVX-Labs/MyPIVXWallet
# [Refactor] Decouple the wallet class

## Abstract

The file `wallet.js` is a real mess: There is no actual wallet manager class and it mixes GUI, crypto functions, ledger stuff and masterKeys. This PR begins a process of refactoring the class:

- A Wallet class has been added that at the moment incorporates all the ugly global wallet related functions that were in `wallet.js`. In the future Wallet will manage UTXOs, balance, addresses, masternodes...
- MasterKey class and derivatives have been moved to a new file `masterkey.js`, masterkeys should have the ONLY role of managing internally the generation of new addresses.
- Hence the global variable `masterKey` has been removed in favor of a global variable of type `Wallet` that has `masterKey` as a private field. At the moment I added a getter to `masterKey` (or the refactor would have been to big), in future PRs we should try to avoid exposing masterkey outside the wallet.
- Ledger related functions and variables have been moved to a new file `ledger.js`
- Finally a new file `encoding.js` has been created. It contains the low-level  cryptographic functions that were previously in the wallet file.

---

## Testing
Test that the wallet functionality are still working as expected for example:

- Create wallet
- Import wallet
- Hit 'Refresh Address' in the Dashboard
- Encrypt (+ Change Password)
- Test Send + Receive
- Test Contacts (Add them, Send to them, etc)


---

## Note
As I already said the refactor is not final and I did not want to change too much lines. In future PRs the getter to `masterKey` and the function `getDerivationPath` should both be removed from the `Wallet` class. We should also really try to decouple the logic from the GUI
